File: ios/Sources/GoogleNavigationPlugin/GoogleNavigationPlugin.swift
import Foundation
import Capacitor

@objc(GoogleNavigationPlugin)
public class GoogleNavigationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GoogleNavigationPlugin"
    public let jsName = "GoogleNavigation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startNavigation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopNavigation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showNavigationView", returnType: CAPPluginReturnPromise),
    ]
    private lazy var implementation = GoogleNavigation(plugin: self)

    @objc func initialize(_ call: CAPPluginCall) {
        let jsKey = call.getString("apiKey") ?? ""
        let resolvedKey = !jsKey.isEmpty
            ? jsKey
            : Bundle.main.object(forInfoDictionaryKey: "GoogleNavigationAPIKey") as? String ?? ""

        guard !resolvedKey.isEmpty else {
            call.reject("No API key found. Add GoogleNavigationAPIKey to Info.plist (recommended) or pass apiKey to initialize().")
            return
        }

        implementation.initialize(apiKey: resolvedKey) { success, error in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve(["success": success])
            }
        }
    }

    @objc func startNavigation(_ call: CAPPluginCall) {
        guard let lat = call.getDouble("destinationLatitude"),
              let lng = call.getDouble("destinationLongitude") else {
            call.reject("destinationLatitude and destinationLongitude are required")
            return
        }
        let travelMode = call.getString("travelMode") ?? "DRIVING"
        implementation.startNavigation(lat: lat, lng: lng, travelMode: travelMode) { success, error in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve(["success": success])
            }
        }
    }

    @objc func stopNavigation(_ call: CAPPluginCall) {
        implementation.stopNavigation()
        call.resolve(["success": true])
    }

    @objc func showNavigationView(_ call: CAPPluginCall) {
        let show = call.getBool("show") ?? true
        DispatchQueue.main.async {
            if show {
                self.implementation.presentNavigationViewController(from: self.bridge?.viewController) { success, error in
                    if let error = error {
                        call.reject(error)
                    } else {
                        call.resolve(["success": success])
                    }
                }
            } else {
                self.implementation.dismissNavigationViewController {
                    call.resolve(["success": true])
                }
            }
        }
    }
}

File: android/src/main/java/com/attributeai/navigation/GoogleNavigationPlugin.java
package com.attributeai.navigation;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "GoogleNavigation")
public class GoogleNavigationPlugin extends Plugin {

    private GoogleNavigation implementation;

    @Override
    public void load() {
        implementation = new GoogleNavigation(this);
    }

    void fireEvent(String eventName, JSObject data) {
        notifyListeners(eventName, data);
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        String apiKey = call.getString("apiKey", "");
        implementation.initialize(apiKey, (success, error) -> {
            if (error != null) {
                call.reject(error);
            } else {
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }
        });
    }

    @PluginMethod
    public void startNavigation(PluginCall call) {
        Double lat = call.getDouble("destinationLatitude");
        Double lng = call.getDouble("destinationLongitude");
        if (lat == null || lng == null) {
            call.reject("destinationLatitude and destinationLongitude are required");
            return;
        }
        String travelMode = call.getString("travelMode", "DRIVING");
        implementation.startNavigation(lat, lng, travelMode, (success, error) -> {
            if (error != null) {
                call.reject(error);
            } else {
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }
        });
    }

    @PluginMethod
    public void stopNavigation(PluginCall call) {
        implementation.stopNavigation();
        JSObject ret = new JSObject();
        ret.put("success", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void showNavigationView(PluginCall call) {
        boolean show = Boolean.TRUE.equals(call.getBoolean("show", true));
        getActivity().runOnUiThread(() -> {
            if (show) {
                implementation.showNavigationView(getActivity(), (success, error) -> {
                    if (error != null) {
                        call.reject(error);
                    } else {
                        JSObject ret = new JSObject();
                        ret.put("success", true);
                        call.resolve(ret);
                    }
                });
            } else {
                implementation.hideNavigationView(getActivity());
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }
        });
    }
}
